config/configparser/parser.go

```diff
@@ -17,96 +17,118 @@ package configparser
 )
 
 // NewParser creates a new empty Parser instance.
 func NewParser() *Parser {
-	return &Parser{
-		v: viper.NewWithOptions(viper.KeyDelimiter(KeyDelimiter)),
-	}
+	return &Parser{k: koanf.New(KeyDelimiter)}
 }
 
 // NewParserFromFile creates a new Parser by reading the given file.
 func NewParserFromFile(fileName string) (*Parser, error) {
 	// Read yaml config from file.
 	p := NewParser()
-	p.v.SetConfigFile(fileName)
-	if err := p.v.ReadInConfig(); err != nil {
+	if err := p.k.Load(file.Provider(fileName), yaml.Parser()); err != nil {
 		return nil, fmt.Errorf("unable to read the file %v: %w", fileName, err)
 	}
 	return p, nil
 }
 
 // NewParserFromBuffer creates a new Parser by reading the given yaml buffer.
 func NewParserFromBuffer(buf io.Reader) (*Parser, error) {
+	content, err := ioutil.ReadAll(buf)
+	if err != nil {
+		return nil, err
+	}
+
 	p := NewParser()
-	p.v.SetConfigType("yaml")
-	if err := p.v.ReadConfig(buf); err != nil {
+	if err := p.k.Load(rawbytes.Provider(content), yaml.Parser()); err != nil {
 		return nil, err
 	}
+
 	return p, nil
 }
 
 // NewParserFromStringMap creates a parser from a map[string]interface{}.
 func NewParserFromStringMap(data map[string]interface{}) *Parser {
 	p := NewParser()
-	// Cannot return error because the viper instance is empty.
-	_ = p.v.MergeConfigMap(data)
+	// Cannot return error because the koanf instance is empty.
+	_ = p.k.Load(confmap.Provider(data, KeyDelimiter), nil)
 	return p
 }
 
 // Parser loads configuration.
 type Parser struct {
-	v *viper.Viper
+	k *koanf.Koanf
 }
 
 // AllKeys returns all keys holding a value, regardless of where they are set.
 // Nested keys are returned with a KeyDelimiter separator.
 func (l *Parser) AllKeys() []string {
-	return l.v.AllKeys()
+	return l.k.Keys()
 }
 
 // Unmarshal unmarshals the config into a struct.
 // Tags on the fields of the structure must be properly set.
 func (l *Parser) Unmarshal(rawVal interface{}) error {
-	return l.v.Unmarshal(rawVal)
+	decoder, err := mapstructure.NewDecoder(decoderConfig(rawVal))
+	if err != nil {
+		return err
+	}
+	return decoder.Decode(l.ToStringMap())
 }
 
 // UnmarshalExact unmarshals the config into a struct, erroring if a field is nonexistent.
 func (l *Parser) UnmarshalExact(intoCfg interface{}) error {
-	return l.v.UnmarshalExact(intoCfg)
+	dc := decoderConfig(intoCfg)
+	dc.ErrorUnused = true
+	decoder, err := mapstructure.NewDecoder(dc)
+	if err != nil {
+		return err
+	}
+	return decoder.Decode(l.ToStringMap())
 }
 
 // Get can retrieve any value given the key to use.
 func (l *Parser) Get(key string) interface{} {
-	return l.v.Get(key)
+	return l.k.Get(key)
 }
 
 // Set sets the value for the key.
 func (l *Parser) Set(key string, value interface{}) {
-	l.v.Set(key, value)
+	// koanf doesn't offer a direct setting mechanism so merging is required.
+	merged := koanf.New(KeyDelimiter)
+	merged.Load(confmap.Provider(map[string]interface{}{key: value}, KeyDelimiter), nil)
+	l.k.Merge(merged)
```

Comment:
Does merge override?

Reply:
Yes: https://github.com/knadh/koanf/blob/a05e2d502ca7d3069ffc14b454956853876e87ab/maps/maps.go#L123.  Here map b is the receiver (our koanf instance) and its values are overwritten by the temporary `merged` map's values.